bfvmm/src/entry/entry.cpp

```diff
@@ -74,42 +78,74 @@ bfobject *
 WEAK_SYM pre_run_vcpu(vcpuid::type id)
 { (void) id; return nullptr; }
 
+bfobject *
+WEAK_SYM pre_hlt_vcpu(vcpuid::type id)
+{ (void) id; return nullptr; }
+
+bfobject *
+WEAK_SYM pre_destroy_vcpu(vcpuid::type id)
+{ (void) id; return nullptr; }
+
+bool
+WEAK_SYM hypervisor_setup(domain_t domain)
+{ bfignored(domain); return true; }
+
+bool
+WEAK_SYM hypervisor_main(vcpu_t vcpu)
+{ bfignored(vcpu); return true; }
+
+bool
+WEAK_SYM hypervisor_teardown(vcpu_t vcpu)
+{ bfignored(vcpu); return true; }
+
 extern "C" int64_t
 private_init_vmm(uint64_t arg) noexcept
 {
     return guard_exceptions(ENTRY_ERROR_VMM_START_FAILED, [&]() {
 
-        g_vcm->create(arg, pre_create_vcpu(arg));
-
-        auto ___ = gsl::on_failure([&]
-        { g_vcm->destroy(arg); });
-
         g_vcm->run(arg, pre_run_vcpu(arg));
 
         return ENTRY_SUCCESS;
     });
 }
 
-bfobject *
-WEAK_SYM pre_hlt_vcpu(vcpuid::type id)
-{ (void) id; return nullptr; }
-
-bfobject *
-WEAK_SYM pre_destroy_vcpu(vcpuid::type id)
-{ (void) id; return nullptr; }
-
 extern "C" int64_t
 private_fini_vmm(uint64_t arg) noexcept
 {
     return guard_exceptions(ENTRY_ERROR_VMM_STOP_FAILED, [&]() {
-
         g_vcm->hlt(arg, pre_hlt_vcpu(arg));
         g_vcm->destroy(arg, pre_destroy_vcpu(arg));
 
         return ENTRY_SUCCESS;
     });
 }
 
+extern "C" bool
+private_hypervisor_setup(uint64_t arg) noexcept
```

Comment:
The main reason I added an extra step here is that I moved all logic that actually touches hardware out of **all** constructors (vmcs, vms, vcpu, etc). This way, all objects that make up a Bareflank hypervisor can be passed an extension's hypervisor_setup function completely constructed, so they can register exit handlers before the vmm is loaded. I agree, there might be a way to do this all in one step, I can look into that